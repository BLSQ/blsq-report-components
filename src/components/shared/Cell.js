import React from "react";
import { withStyles } from "@material-ui/core/styles";
import {
  numberWithCommas,
  roundedAmount,
  roundedAmountOrInteger,
  roundedPercent,
  labelize
} from "../../support/NumberFormatter";

const styles = theme => ({
  cellOrder: {
    textAlign: "center",
    padding: "0px 3px 0px 3px",
    border: "0.5pt solid black"
  },
  cellText: {
    textAlign: "left",
    height: "21px",
    padding: "0px 3px 0px 3px",
    border: "0.5pt solid black"
  },
  cellLargeText: {
    textAlign: "left",
    height: "18px",
    padding: "5px 5px 5px 5px",
    border: "0.5pt solid black"
  },
  cellQuantity: {
    textAlign: "center",
    padding: "0px 3px 0px 3px",
    border: "0.5pt solid black"
  },
  cellAmount: {
    textAlign: "right",
    padding: "0px 3px 0px 3px",
    border: "0.5pt solid black"
  }
});

function resolve(path, obj, separator = ".") {
  if (path === "self") {
    return obj;
  }
  var properties = Array.isArray(path) ? path : path.split(separator);
  return properties.reduce((prev, curr) => prev && prev[curr], obj);
}

const Cell = props => {
  const { classes, variant, field, value, bold, ...other } = props;
  const amount = resolve(field, value);
  let className = null;
  let label = labelize(amount);
  let displayedValue = amount;

  if (variant === undefined || variant === "quantity") {
    if (amount === undefined) {
      displayedValue = "";
    } else {
      displayedValue = amount.value;
    }
    className = classes.cellQuantity;
  } else if (variant === "money") {
    displayedValue = numberWithCommas(roundedAmount(amount.value));
    className = classes.cellAmount;
  } else if (variant === "rounded") {
    displayedValue = roundedAmount(amount.value);
    className = classes.cellQuantity;
  } else if (variant === "text") {
    label = "";
    displayedValue = amount;
    className = classes.cellText;
  } else if (variant === "largeText") {
    label = "";
    displayedValue = amount;
    className = classes.cellLargeText;
  } else if (variant === "textde") {
    if (amount !== undefined) {
      displayedValue = amount.value;
    }
    className = classes.cellText;
  } else if (variant === "order") {
    label = "";
    displayedValue = amount;
    className = classes.cellOrder;
  } else if (props.variant === "percentage") {
    displayedValue = roundedPercent(amount.value);
    className = classes.cellQuantity;
  } else if (props.variant === "roundedAmountOrInteger") {
    displayedValue = roundedAmountOrInteger(amount.value);
    className = classes.cellQuantity;
  } else {
    throw new Error("not supported variant : " + variant);
  }

  const boldValue = bold === true;
  return (
    <td className={className} title={label} {...other}>
      {boldValue && <b>{displayedValue}</b>}
      {!boldValue && <React.Fragment>{displayedValue}</React.Fragment>}
    </td>
  );
};

export default withStyles(styles)(Cell);
