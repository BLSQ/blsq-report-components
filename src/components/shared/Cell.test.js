import React from "react";
import Cell from "./Cell";
import TestRenderer from "react-test-renderer";

describe("Cell", () => {
  const value = {
    demo: {
      code: "demo code",
      value: 15564.524,
      name: "super",
      period: "2016Q3"
    }
  };

  it("is truthy", () => {
    expect(Cell).toBeTruthy();
  });

  it("renders text with self field", () => {
    const tree = TestRenderer.create(
      <Cell variant="text" field="self" value="demo text" />
    );

    expect(tree.toJSON()).toMatchSnapshot();
  });

  it("renders amount with provided field", () => {
    const tree = TestRenderer.create(
      <Cell variant="money" field="demo" value={value} />
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it("renders percentage with provided field", () => {
    const tree = TestRenderer.create(
      <Cell variant="percentage" field="demo" value={value} bold />
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it("renders rounded percentage with provided field", () => {
    const tree = TestRenderer.create(
      <Cell variant="roundedPercentage" field="demo" value={value} bold />
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it("renders title with self field", () => {
    const tree = TestRenderer.create(
      <Cell variant="title" field="self" value="demo title" />
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it("renders text from nested field with link", () => {
    const tree = TestRenderer.create(
      <Cell
        variant="text"
        field="demo.name"
        value={value}
        href="https://google.com"
        bold
      />
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
