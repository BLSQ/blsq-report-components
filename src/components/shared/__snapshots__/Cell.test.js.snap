// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell renders amount with provided field 1`] = `
<td
  className="Cell-cellAmount-5"
  title="super (demo code)  2016Q3 15564.524"
>
  15,564.52
</td>
`;

exports[`Cell renders percentage with provided field 1`] = `
<td
  className="Cell-cellQuantity-4"
  title="super (demo code)  2016Q3 15564.524"
>
  <b>
    15564.52
  </b>
</td>
`;

exports[`Cell renders text from nested field with link 1`] = `
<td
  className="Cell-cellText-2"
  title=""
>
  <a
    className="Cell-customLink-6"
    href="https://google.com"
  >
    <b>
      super
    </b>
  </a>
</td>
`;

exports[`Cell renders text with self field 1`] = `
<td
  className="Cell-cellText-2"
  title=""
>
  demo text
</td>
`;
